docs/proposals/TrafficMirror.md

```diff
@@ -0,0 +1,46 @@
+# Traffic Mirroring
+
+## Context
+
+Traffic Mirroring is a useful feature which sends a copy of live traffic to the mirrored service which will help in testing a service with real production traffic.The difference is that the reply from the mirrored service is dropped by the envoy proxy and not returned to the caller service.
+
+	    a                      b                     c
+	|----------|             |-----------|    mirror   |------------|
+	| service A| ----------> | Service B | ----------> |  Service C |
+	|----------|             |-----------|             |------------|
+
+Envoy provides full support of traffic mirroring [Envoy's Traffic Mirror](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route_components.proto#envoy-api-msg-route-routeaction-requestmirrorpolicy),
+so to support the same functionality in kuma.
+
+
+## Proposed configuration model
+
+We need to extend `TrafficRoute` policy as below
+
+```yaml
+type: TrafficRoute
+mesh: default
+metadata:
+  name: route-example
+spec:
+  sources:
+    - match:
+        kuma.io/service: backend_default_svc_80
+  destinations:
+    - match:
+        kuma.io/service: redis_default_svc_6379
+  conf:
+    split:
+      - weight: 100
+        destination:
+          kuma.io/service: redis_default_svc_6379
+          version: '1.0'
+        mirror:
+          weight: 50 # percentage of traffic to be sent to the mirrored service.
+          destination:
+            kuma.io/service: redis_default_svc_6379
+            version: '2.0'
```

Comment:
What's the rationale between having this inline rather than at the split level?

Like:
```yaml
conf:
  split:
    - weight: 100
      destination:
           kuma.io/service: redis_default_svc_6379
           version: '1.0'
  mirror:
     - weight: 10
       destination:
           kuma.io/service: redis_default_svc_6379
           version: "2.0"
```

Istio does this at this level as well: https://istio.io/latest/docs/tasks/traffic-management/mirroring/

Also should it be a percent instead of a weight? Might be more straight forward for users.

Reply:
I was trying to give a better name and used weight as from the route config. how about `mirrorPercent` or `percentage` ?

Reply:
I like `percentage` as it's already in a `mirror` section `mirrorPercent` seems redundant.

Reply:
So, IMO as a regular use case, we will be normally mirroring the request to a single service and not many. So decided to kept it as an inline. But, agreed with this should be at the split level as envoy supports mirroring to multiple hosts (https://github.com/envoyproxy/envoy/pull/9164). But, how about keeping the mirror as an array inside the split as an inline ( split traffic to 2 services and these 2 services can have the ability to mirror to multiple services)?

Reply:
@lahabana also if we keep the configuration for at split level means, it will not be a good UX, if it is inline means the user can be able to understand we are mirroring the routed service traffic.

Reply:
The reason for this is, In envoy mirroring configuration is at the `route` level policy and not at the cluster level.